homeassistant/components/saj/sensor.py

```diff
@@ -48,139 +58,161 @@
-            )
 
-    async_add_entities(hass_sensors)
+def _init_pysaj(wifi, config):  # pragma: no cover
+    kwargs = {"wifi": wifi}
+    if config.get(CONF_USERNAME) and config.get(CONF_PASSWORD):
+        kwargs["username"] = config[CONF_USERNAME]
+        kwargs["password"] = config[CONF_PASSWORD]
 
-    async def async_saj():
-        """Update all the SAJ sensors."""
-        values = await saj.read(sensor_def)
+    return pysaj.SAJ(config[CONF_HOST], **kwargs)
 
-        for sensor in hass_sensors:
-            state_unknown = False
-            # SAJ inverters are powered by DC via solar panels and thus are
-            # offline after the sun has set. If a sensor resets on a daily
-            # basis like "today_yield", this reset won't happen automatically.
-            # Code below checks if today > day when sensor was last updated
-            # and if so: set state to None.
-            # Sensors with live values like "temperature" or "current_power"
-            # will also be reset to None.
-            if not values and (
-                (sensor.per_day_basis and date.today() > sensor.date_updated)
-                or (not sensor.per_day_basis and not sensor.per_total_basis)
-            ):
-                state_unknown = True
-            sensor.async_update_values(unknown_state=state_unknown)
 
-        return values
+class SAJInverter:
+    """Representation of a SAJ inverter."""
+
+    def __init__(self, config, saj=None):
+        """Init SAJ Inverter class."""
+        self._name = config.get(CONF_NAME)
+        wifi = config[CONF_TYPE] == INVERTER_TYPES[1]
 
-    def start_update_interval(event):
-        """Start the update interval scheduling."""
-        nonlocal remove_interval_update
-        remove_interval_update = async_track_time_interval_backoff(hass, async_saj)
+        self._saj = saj or _init_pysaj(wifi, config)
+        self._sensor_def = pysaj.Sensors(wifi)
+        if CONF_DEVICE_ID in config:
+            self._saj.serialnumber = config[CONF_DEVICE_ID]
+        if ENABLED_SENSORS in config:
+            for sensor in self._sensor_def:
+                sensor.enabled = sensor.key in config[ENABLED_SENSORS]
 
-    def stop_update_interval(event):
-        """Properly cancel the scheduled update."""
-        remove_interval_update()  # pylint: disable=not-callable
+        self.coordinator = None
+        self._interval = config.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
 
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, start_update_interval)
-    hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, stop_update_interval)
+    def get_enabled_sensors(self):
+        """Return enabled sensors keys."""
+        return [s.key for s in self._sensor_def if s.enabled]
 
+    @property
+    def available(self) -> bool:
+        """Return True if device is available."""
+        return self.coordinator and self.coordinator.last_update_success
```

Comment:
This is handled on the entity level by the CoordinatorEntity, it isn't used in this class.